### ISSUE TITLE
Error deserializing object of type unsigned long  while deserializing object of type std::vector

### ISSUE THREAD

**Issue by zqr688** [davisking/dlib#737]
Hello,I use **dnn_imagenet_train_ex.cpp** sample training picture classification, And get the result ‘’image.dnn“.  But the  **dnn_imagenet_ex.cpp** sample uses this model to get an error .

The error is as follows:
**Error deserializing object of type unsigned long
   while deserializing object of type std::vector**

The code is as follows

```
int main(int argc, char** argv)try {
	string imagenet_root_dir = "/opt/image_en/train";
	auto listing = get_imagenet_train_listing(imagenet_root_dir);
	cout << "images in dataset: " << listing.size() << endl;
	const auto number_of_classes = listing.back().numeric_label + 1;
	if (listing.size() == 0)
	{
		cout << "Didn't find the imagenet dataset. " << endl;
		return 1;
	}


	set_dnn_prefer_smallest_algorithms();


	const double initial_learning_rate = 0.1;
	const double weight_decay = 0.0001;
	const double momentum = 0.9;

	net_type net;
	dnn_trainer<net_type> trainer(net, sgd(weight_decay, momentum));
	trainer.be_verbose();
	trainer.set_learning_rate(initial_learning_rate);
	trainer.set_synchronization_file("imagenet_trainer_state_file.dat", std::chrono::minutes(10));
	// This threshold is probably excessively large.  You could likely get good results
	// with a smaller value but if you aren't in a hurry this value will surely work well.
	trainer.set_iterations_without_progress_threshold(5000);
	// Since the progress threshold is so large might as well set the batch normalization
	// stats window to something big too.
	set_all_bn_running_stats_window_sizes(net, 250);

	std::vector<matrix<rgb_pixel>> samples;
	std::vector<unsigned long> labels;
	// Start a bunch of threads that read images from disk and pull out random crops.  It's
	// important to be sure to feed the GPU fast enough to keep it busy.  Using multiple
	// thread for this kind of data preparation helps us do that.  Each thread puts the
	// crops into the data queue.
	dlib::pipe<std::pair<image_info, matrix<rgb_pixel>>> data(200);
	auto f = [&data, &listing](time_t seed)
	{
		dlib::rand rnd(time(0) + seed);
		matrix<rgb_pixel> img;
		std::pair<image_info, matrix<rgb_pixel>> temp;
		while (data.is_enabled())
		{
			temp.first = listing[rnd.get_random_32bit_number() % listing.size()];
			load_image(img, temp.first.filename);
			randomly_crop_image(img, temp.second, rnd);
			data.enqueue(temp);
		}
	};
	std::thread data_loader1([f]() { f(1); });
	std::thread data_loader2([f]() { f(2); });
	std::thread data_loader3([f]() { f(3); });
	std::thread data_loader4([f]() { f(4); });
	// The main training loop.  Keep making mini-batches and giving them to the trainer.
	// We will run until the learning rate has dropped by a factor of 1e-3.
	//while (trainer.get_learning_rate() >= initial_learning_rate*1e-1)
	int k=0;
	while (k<6)
	{
		samples.clear();
		labels.clear();
		// make a 160 image mini-batch
		std::pair<image_info, matrix<rgb_pixel>> img;
		while (samples.size() < 160)
		{
			data.dequeue(img);
			samples.push_back(std::move(img.second));
			labels.push_back(img.first.numeric_label);
		}
		trainer.train_one_step(samples, labels);
		k++;
	}
	// Training done, tell threads to stop and make sure to wait for them to finish before
	// moving on.
	data.disable();
	data_loader1.join();
	data_loader2.join();
	data_loader3.join();
	data_loader4.join();

	// also wait for threaded processing to stop in the trainer.
	trainer.get_net();
	net.clean();
	cout << "saving network" << endl;
	serialize(imagenet_root_dir + "/image.dnn") << net;
}
catch (std::exception& e)
{
	cout << e.what() << endl;
}

```

I just want to verify whether the training can be successful, so iterated only 6 times.

How to deal with this problem, thanks a lot.

**Comment by davisking** [comment]
The code looks fine.  You probably just typed the filename wrong into your other program.

**Comment by zqr688** [comment]
The other program is as follows:
'''int main(int argc, char** argv) try
{
   
    string imagepath = "/opt/image_en/train/cat/cat_1.jpg";
    std::vector<string> labels;
    anet_type net;
    deserialize("/opt/image_en/train/image.dnn") >> net >> labels;

    // Make a network with softmax as the final layer.  We don't have to do this
    // if we just want to output the single best prediction, since the anet_type
    // already does this.  But if we instead want to get the probability of each
    // class as output we need to replace the last layer of the network with a
    // softmax layer, which we do as follows:
    softmax<anet_type::subnet_type> snet; 
    snet.subnet() = net.subnet();

    dlib::array<matrix<rgb_pixel>> images;
    matrix<rgb_pixel> img, crop;

    dlib::rand rnd;
    image_window win;

    // Read images from the command prompt and print the top 5 best labels for each.
     load_image(img, imagepath );
     const int num_crops = 16;
     // Grab 16 random crops from the image.  We will run all of them through the
     // network and average the results.
     randomly_crop_images(img, images, rnd, num_crops);
     // p(i) == the probability the image contains object of class i.
     matrix<float,1,1000> p = sum_rows(mat(snet(images.begin(), images.end())))/num_crops;

        // Print the 5 most probable labels
     for (int k = 0; k < 5; ++k)
     {
         unsigned long predicted_label = index_of_max(p);
         cout << p(predicted_label) << ": " << labels[predicted_label] << endl;
         p(predicted_label) = 0;
     }   
  
}
catch(std::exception& e)
{
    cout << e.what() << endl;
}'''

If "image.dnn" is replaced with “resnet34_1000_imagenet_classifier.dnn”, the result can be returned normally.

I do not know why that happened. Can you help me. Many thanks.

**Comment by omarkacem** [comment]
Hello,I use dnn_imagenet_train_ex.cpp sample training picture classification, And get the result ‘’image.dnn“. But the dnn_imagenet_ex.cpp give me the same error:
The error is as follows:
Error deserializing object of type unsigned long
while deserializing object of type std::vector.

I do not know why that happened. Can you help me.Thanks.

**Comment by omarkacem** [comment]
hello zqr688,are you fixed this error or not yet. please help me and thank you .

**Comment by omarkacem** [comment]
hello zqr688,did you fixed this error or not yet. please help me and thank you .

**Comment by omarkacem** [comment]
Hello Mr @davisking,I have a problem and can you help me please :
- I want to classify my own images and i create my imagenet_train_images.txt file like yours and get the result 'image.dnn" but the dnn_imagenet_ex.cpp give me the same error:
Error deserializing object of type unsigned long
while deserializing object of type std::vector.

Can you help me if i forget something.
I do not know why that happened. Can you help me.

**Comment by omarkacem** [comment]
hello @zqr688 ,did you fixed this error or not yet. please help me and thank you.
